[CSS Zoom] Apply zoom factor to LineWidth keyword values

nullhook opened the pull request:
#### e9758b51c95e9d00a12943641c22248ba004841a
<pre>
[CSS Zoom] Apply zoom factor to LineWidth keyword values
<a href="https://bugs.webkit.org/show_bug.cgi?id=302111">https://bugs.webkit.org/show_bug.cgi?id=302111</a>
<a href="https://rdar.apple.com/164193590">rdar://164193590</a>

Reviewed by Sam Weinig.

We apply the zoom factor to LineWidth keyword values because it
follows the correct spec behavior. Effectively this also removes branching
in the Evaluation&lt;T&gt; specialization and reduces the struct size.

Tests: fast/css/border-width-keywords-non-eval-zoom.html
       fast/css/column-rule-width-keywords-non-eval-zoom.html
       fast/css/outline-width-keywords-non-eval-zoom.html
       imported/w3c/web-platform-tests/css/css-viewport/zoom/border-width-keywords.html
       imported/w3c/web-platform-tests/css/css-viewport/zoom/column-rule-width-keywords.html
       imported/w3c/web-platform-tests/css/css-viewport/zoom/outline-width-keywords.html
       imported/w3c/web-platform-tests/css/css-viewport/zoom/reference/border-width-keywords-ref.html
       imported/w3c/web-platform-tests/css/css-viewport/zoom/reference/column-rule-width-keywords-ref.html
       imported/w3c/web-platform-tests/css/css-viewport/zoom/reference/outline-width-keywords-ref.html
* LayoutTests/fast/backgrounds/size/contain-and-cover-zoomed-expected.txt:

* LayoutTests/fast/canvas/canvas-zoom-expected.html:

Now that zoom is applied to keywords, the reference has been updated
to use an explicit 6px border (3px medium * zoom 2) instead of
relying on the unspecified border-width value.

* LayoutTests/fast/css/border-width-keywords-non-eval-zoom-expected.html: Added.
* LayoutTests/fast/css/border-width-keywords-non-eval-zoom.html: Added.
* LayoutTests/fast/css/column-rule-width-keywords-non-eval-zoom-expected.html: Added.
* LayoutTests/fast/css/column-rule-width-keywords-non-eval-zoom.html: Added.
* LayoutTests/fast/css/outline-width-keywords-non-eval-zoom-expected.html: Added.
* LayoutTests/fast/css/outline-width-keywords-non-eval-zoom.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-viewport/zoom/border-width-keywords-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-viewport/zoom/border-width-keywords.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-viewport/zoom/column-rule-width-keywords-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-viewport/zoom/column-rule-width-keywords.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-viewport/zoom/outline-width-keywords-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-viewport/zoom/outline-width-keywords.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-viewport/zoom/reference/border-width-keywords-ref.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-viewport/zoom/reference/column-rule-width-keywords-ref.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-viewport/zoom/reference/outline-width-keywords-ref.html: Added.
* LayoutTests/platform/ios/fast/backgrounds/background-position-parsing-expected.txt:
* LayoutTests/platform/mac/fast/backgrounds/background-position-parsing-expected.txt:
* LayoutTests/platform/wpe-wk2/fast/backgrounds/background-position-parsing-expected.txt: Added.
* LayoutTests/platform/wpe/fast/backgrounds/background-position-parsing-expected.txt:
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/rendering/style/BorderValue.h:
* Source/WebCore/rendering/style/OutlineValue.h:
* Source/WebCore/rendering/style/RenderStyle.cpp:
* Source/WebCore/rendering/style/RenderStyleInlines.h:
(WebCore::RenderStyle::initialBorderWidth):
(WebCore::RenderStyle::initialColumnRuleWidth):
(WebCore::RenderStyle::initialOutlineWidth):
* Source/WebCore/style/StyleBuilderCustom.h:
(WebCore::Style::BuilderCustom::applyInitialBorderTopWidth):
(WebCore::Style::BuilderCustom::applyInitialBorderRightWidth):
(WebCore::Style::BuilderCustom::applyInitialBorderBottomWidth):
(WebCore::Style::BuilderCustom::applyInitialBorderLeftWidth):
(WebCore::Style::BuilderCustom::applyInitialOutlineWidth):
(WebCore::Style::BuilderCustom::applyInitialColumnRuleWidth):
* Source/WebCore/style/values/backgrounds/StyleLineWidth.cpp:
(WebCore::Style::handleKeywordValue):
(WebCore::Style::CSSValueConversion&lt;LineWidth&gt;::operator):
* Source/WebCore/style/values/backgrounds/StyleLineWidth.h:
(WebCore::Style::LineWidth::LineWidth):

Canonical link: <a href="https://commits.webkit.org/303044@main">https://commits.webkit.org/303044@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/07a25ae62ee620c47837741cf48764c487c301cb

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/130857 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/3180 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/41816 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/138285 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/82516 "Built successfully") | ⏳ 🛠 ios-apple 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/132728 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/3162 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/3023 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/99703 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/67527 "Passed tests") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/ac68a553-32dd-4270-abe6-6d9e3b1bc6f2) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/133803 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/2274 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/117175 "Passed tests") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/80397 "Found 1 new API test failure: WPE/TestWebsiteData:/webkit/WebKitWebsiteData/itp (failure)") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/cc3bed35-883e-454c-9576-e823b0c4d765) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/2194 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/35303 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/81540 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/110787 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/35806 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/140762 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/2924 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/2633 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/108218 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/2970 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/113506 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/108138 "Passed tests") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/2237 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/31932 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/55958 "Built successfully") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/20395 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/2992 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/66384 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/2813 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/3013 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/2921 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/style/values/backgrounds/StyleLineWidth.cpp:
```diff
@@ -37,25 +37,39 @@ namespace Style {
 
+static auto handleKeywordValue(BuilderState& state, CSSValueID valueID) -> LineWidth
+{
+    float keywordValue;
+    switch (valueID) {
+    case CSSValueThin:
+        keywordValue = 1.0f;
+        break;
+    case CSSValueMedium:
+        keywordValue = 3.0f;
+        break;
+    case CSSValueThick:
+        keywordValue = 5.0f;
+        break;
+    default:
+        state.setCurrentPropertyInvalidAtComputedValueTime();
+        keywordValue = 3.0f; // Medium
```
Could we somehow not duplicate this with the values in the header file?

nullhook replied:
there is a way, but it creates a temp object on the stack and calls unresolvedValue() to extract the float, which may introduce an overhead.

as we saw in some workloads, these paths are hot. i'm trying to avoid causing another regression. the primary goal for this PR is correctness, but also to eliminate branching/overhead and ideally see some progression.

```
static auto handleKeywordValue(...)
{
      LineWidth unzoomed(..);
      switch (valueID) {
      case CSSValueThin:
          unzoomed = CSS::Keyword::Thin { };
          break;
       ...
       
       float keywordValue = unzoomed.value.unresolvedValue();
       return LineWidth::Length { floorToDevicePixel(keywordValue * state.style().usedZoom(), state.document().deviceScaleFactor()) };
 }         
```

nullhook replied:
(which reminds me if i should annotate handleKeywordValue with inline)

weinig commented on LayoutTests/fast/backgrounds/size/contain-and-cover-zoomed-expected.txt:
```diff
@@ -1,13 +1,13 @@
 layer at (0,0) size 800x600
```
How much effort would it be to turn this into a ref test?

nullhook replied:
converting this to a ref-test seems non-trivial because the page zoom affects many properties (width, height, border, margin). we'll have to manually compute all the zoomed properties in the reference html

weinig commented on LayoutTests/fast/canvas/canvas-zoom-expected.html:
```diff
@@ -1,5 +1,5 @@
 <style>
-    canvas { border: solid green; }
+    canvas { border: 6px solid green; }
```
What is this change for? The commit message should probably explain the changes to these tests.

nullhook replied:
added an explanation

weinig commented on LayoutTests/imported/w3c/web-platform-tests/css/css-viewport/zoom/border-width-keywords.html:
```diff
@@ -0,0 +1,35 @@
+}
+</style>
+
+<div class="thin"></div>
```
I think these tests usually try to have body, head and html tags.

weinig commented on Source/WebCore/style/values/backgrounds/StyleLineWidth.cpp:
```diff
@@ -37,25 +37,39 @@ namespace Style {
 
 // MARK: - Conversion
 
+static inline auto handleKeywordValue(BuilderState& state, CSSValueID valueID) -> LineWidth
```
No need for `inline` here.

weinig commented on Source/WebCore/style/values/backgrounds/StyleLineWidth.h:
```diff
@@ -43,14 +43,13 @@ struct LineWidth {
-    constexpr LineWidth(CSS::Keyword::Thin) : value { 1 }, isKeyword { true } { }
-    constexpr LineWidth(CSS::Keyword::Medium) : value { 3 }, isKeyword { true } { }
-    constexpr LineWidth(CSS::Keyword::Thick) : value { 5 }, isKeyword { true } { }
+    constexpr LineWidth(CSS::Keyword::Thin) : value { 1 } { }
```
I think it probably makes sense to remove these constructors taking keywords for now, as a caller is probably doing the wrong thing.

nullhook replied:
why do you think the callers are doing the wrong thing here?

i do see callers like `Style::LineWidth m_width { CSS::Keyword::Medium { } };` in OutlineValue.h that need a default. there are other codepaths that return these keywords simply too.

weinig replied:
I think its wrong, because now when one would use a keyword, zoom should also be applied.

weinig reviewed (changes requested):
It seems like this probably is broken for something like

```
zoom: 2;
border-bottom-width: initial;
```

with zoom, since `RenderStyle::initialBorderWidth()` will create a LineWidth without applying zoom.

weinig commented on Source/WebCore/rendering/style/RenderStyleInlines.h:
```diff
@@ -397,7 +397,7 @@ constexpr Style::ColumnCount RenderStyle::initialColumnCount() { return CSS::Key
 inline Style::GapGutter RenderStyle::initialColumnGap() { return CSS::Keyword::Normal { }; }
 constexpr ColumnProgression RenderStyle::initialColumnProgression() { return ColumnProgression::Normal; }
-constexpr Style::LineWidth RenderStyle::initialColumnRuleWidth() { return CSS::Keyword::Medium { }; }
+constexpr Style::LineWidth RenderStyle::initialColumnRuleWidth() { return Style::LineWidth { 3.0f }; }
```
This still seems like it will be wrong, since no zoom will be applied. I think you need to use a "style-builder-custom": "Initial" for these properties and and ensure the zoom is applied.

weinig replied:
To test this you will need two tests cases, one in the new mode, and one with the new mode explicitly disabled.

nullhook replied:
so, effectively the generated code would call applyInitialProp here which we need to create a custom handler for?

i can technically do a `Style::evaluate` with `usedZoomForLength` here but trying to avoid that in non-evaluation mode.

```
inline void BuilderCustom::applyInitialBorderTopWidth(BuilderState& builderState)
{
    if (evaluationTimeZoomEnabled(builderState))
        builderState.style().setBorderTopWidth(RenderStyle::initialBorderWidth());
    else
        builderState.style().setBorderTopWidth(Style::LineWidth { RenderStyle::initialBorderWidth().value.unresolvedValue() * builderState.style().usedZoom() });
}
```

weinig commented on LayoutTests/fast/css/outline-width-keywords-non-eval-zoom.html:
```diff
@@ -0,0 +1,53 @@
+<!-- webkit-test-runner [ EvaluationTimeZoomEnabled=false ] -->
```
This one, with EvaluationTimeZoomEnabled=false, should not be added to wpt, so I would put this in fast/ somewhere instead.

weinig commented on Source/WebCore/style/StyleBuilderCustom.h:
```diff
@@ -182,6 +183,12 @@ class BuilderCustom {
 public:
     // Custom handling of inherit, initial and value setting.
     // FIXME: <https://webkit.org/b/212506> Teach makeprop.pl to generate setters for hasExplicitlySet* flags
+    DECLARE_PROPERTY_CUSTOM_HANDLERS(BorderTopWidth);
```
DECLARE_PROPERTY_CUSTOM_HANDLERS will add all three (Initial, Inherit and Value). You just want Initial for these, so can just declare them normally like:

```c++
    // Custom handling of initial setting only.
    static void applyInheritBorderTopWidth(BuilderState&);
    static void applyInheritBorderRightWidth(BuilderState&);
    static void applyInheritBorderBottomWidth(BuilderState&);
    static void applyInheritBorderLeftWidth(BuilderState&);
    static void applyInheritOutlineWidth(BuilderState&);
    static void applyInheritColumnRuleWidth(BuilderState&);
```

nullhook committed:
[CSS Zoom] Apply zoom factor to LineWidth keyword values
https://bugs.webkit.org/show_bug.cgi?id=302111
rdar://164193590

Reviewed by Sam Weinig.

We apply the zoom factor to LineWidth keyword values because it
follows the correct spec behavior. Effectively this also removes branching
in the Evaluation<T> specialization and reduces the struct size.

Tests: fast/css/border-width-keywords-non-eval-zoom.html
       fast/css/column-rule-width-keywords-non-eval-zoom.html
       fast/css/outline-width-keywords-non-eval-zoom.html
       imported/w3c/web-platform-tests/css/css-viewport/zoom/border-width-keywords.html
       imported/w3c/web-platform-tests/css/css-viewport/zoom/column-rule-width-keywords.html
       imported/w3c/web-platform-tests/css/css-viewport/zoom/outline-width-keywords.html
       imported/w3c/web-platform-tests/css/css-viewport/zoom/reference/border-width-keywords-ref.html
       imported/w3c/web-platform-tests/css/css-viewport/zoom/reference/column-rule-width-keywords-ref.html
       imported/w3c/web-platform-tests/css/css-viewport/zoom/reference/outline-width-keywords-ref.html
* LayoutTests/fast/backgrounds/size/contain-and-cover-zoomed-expected.txt:

* LayoutTests/fast/canvas/canvas-zoom-expected.html:

Now that zoom is applied to keywords, the reference has been updated
to use an explicit 6px border (3px medium * zoom 2) instead of
relying on the unspecified border-width value.

* LayoutTests/fast/css/border-width-keywords-non-eval-zoom-expected.html: Added.
* LayoutTests/fast/css/border-width-keywords-non-eval-zoom.html: Added.
* LayoutTests/fast/css/column-rule-width-keywords-non-eval-zoom-expected.html: Added.
* LayoutTests/fast/css/column-rule-width-keywords-non-eval-zoom.html: Added.
* LayoutTests/fast/css/outline-width-keywords-non-eval-zoom-expected.html: Added.
* LayoutTests/fast/css/outline-width-keywords-non-eval-zoom.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-viewport/zoom/border-width-keywords-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-viewport/zoom/border-width-keywords.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-viewport/zoom/column-rule-width-keywords-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-viewport/zoom/column-rule-width-keywords.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-viewport/zoom/outline-width-keywords-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-viewport/zoom/outline-width-keywords.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-viewport/zoom/reference/border-width-keywords-ref.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-viewport/zoom/reference/column-rule-width-keywords-ref.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-viewport/zoom/reference/outline-width-keywords-ref.html: Added.
* LayoutTests/platform/ios/fast/backgrounds/background-position-parsing-expected.txt:
* LayoutTests/platform/mac/fast/backgrounds/background-position-parsing-expected.txt:
* LayoutTests/platform/wpe-wk2/fast/backgrounds/background-position-parsing-expected.txt: Added.
* LayoutTests/platform/wpe/fast/backgrounds/background-position-parsing-expected.txt:
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/rendering/style/BorderValue.h:
* Source/WebCore/rendering/style/OutlineValue.h:
* Source/WebCore/rendering/style/RenderStyle.cpp:
* Source/WebCore/rendering/style/RenderStyleInlines.h:
(WebCore::RenderStyle::initialBorderWidth):
(WebCore::RenderStyle::initialColumnRuleWidth):
(WebCore::RenderStyle::initialOutlineWidth):
* Source/WebCore/style/StyleBuilderCustom.h:
(WebCore::Style::BuilderCustom::applyInitialBorderTopWidth):
(WebCore::Style::BuilderCustom::applyInitialBorderRightWidth):
(WebCore::Style::BuilderCustom::applyInitialBorderBottomWidth):
(WebCore::Style::BuilderCustom::applyInitialBorderLeftWidth):
(WebCore::Style::BuilderCustom::applyInitialOutlineWidth):
(WebCore::Style::BuilderCustom::applyInitialColumnRuleWidth):
* Source/WebCore/style/values/backgrounds/StyleLineWidth.cpp:
(WebCore::Style::handleKeywordValue):
(WebCore::Style::CSSValueConversion<LineWidth>::operator):
* Source/WebCore/style/values/backgrounds/StyleLineWidth.h:
(WebCore::Style::LineWidth::LineWidth):

Canonical link: https://commits.webkit.org/303044@main
